packages/@lwc/template-compiler/src/parser/expression.ts

```diff
@@ -33,57 +33,62 @@ export function isPotentialExpression(source: string): boolean {
+    invariant(!hasMultipleExpressions, ParserDiagnostics.MULTIPLE_EXPRESSIONS);
+
+    const expressionStatement = (ast as estree.Program).body[0];
+    invariant(expressionStatement.type === 'ExpressionStatement', ParserDiagnostics.INVALID_NODE, [
+        expressionStatement.type,
+    ]);
+
+    const expression = (expressionStatement as estree.ExpressionStatement).expression;
+
+    validateExpression(expression, element, state.config.experimentalComputedMemberExpression);
+
+    return (expression as unknown) as TemplateExpression;
+}
+
 // FIXME: Avoid throwing errors and return it properly
 export function parseExpression(source: string, element: IRNode, state: State): TemplateExpression {
     try {
-        const parsed = babylon.parse(source);
-
-        let expression: any;
-
-        traverse(parsed, {
-            enter(path) {
-                const isValidNode =
-                    path.isProgram() ||
-                    path.isBlockStatement() ||
-                    path.isExpressionStatement() ||
-                    path.isIdentifier() ||
-                    path.isMemberExpression();
-                invariant(isValidNode, ParserDiagnostics.INVALID_NODE, [path.type]);
-
-                // Ensure expression doesn't contain multiple expressions: {foo;bar}
-                const hasMultipleExpressions =
-                    path.isBlock() && (path.get('body') as any).length !== 1;
-                invariant(!hasMultipleExpressions, ParserDiagnostics.MULTIPLE_EXPRESSIONS);
-
-                // Retrieve the first expression and set it as return value
-                if (path.isExpressionStatement() && !expression) {
-                    expression = (path.node as types.ExpressionStatement).expression;
-                }
-            },
-
-            MemberExpression: {
-                exit(path) {
-                    const shouldReportComputed =
-                        !state.config.experimentalComputedMemberExpression &&
-                        (path.node as types.MemberExpression).computed;
-                    invariant(
-                        !shouldReportComputed,
-                        ParserDiagnostics.COMPUTED_PROPERTY_ACCESS_NOT_ALLOWED
-                    );
-
-                    const memberExpression = path.node as types.MemberExpression;
-                    const propertyIdentifier = memberExpression.property as TemplateIdentifier;
-                    const objectIdentifier = memberExpression.object as TemplateIdentifier;
-                    invariant(
-                        !isBoundToIterator(objectIdentifier, element) ||
-                            propertyIdentifier.name !== ITERATOR_NEXT_KEY,
-                        ParserDiagnostics.MODIFYING_ITERATORS_NOT_ALLOWED
-                    );
-                },
-            },
-        });
+        const parsed = parse(source.substr(1, source.length - 2), { ecmaVersion: 2020 });
```

Comment:
The `parse` method does a full module parsing. For performance purposes, I think we should rather use the `parseExpressionAt` method. If you look at acorn source code, you can see that it skips all the module parsing and jump directly to expression parsing which is nice to have performance optimization.

Something worth mentioning is that with the `parseExpressionAt` is that acorn will stop at the end of the expression, so we will have to check that expression is only followed with whitespaces. For reference here is how Svelte compiler is using the [parseExpressionAt](https://github.com/sveltejs/svelte/blob/master/src/compiler/parse/read/expression.ts) method.